core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinFastRecordCursorFactory.java

```diff
@@ -0,0 +1,460 @@
+import io.questdb.griffin.engine.LimitOverflowException;
+import io.questdb.griffin.model.JoinContext;
+import io.questdb.std.*;
+import io.questdb.std.str.Utf8Sequence;
+
+public final class AsOfJoinFastRecordCursorFactory extends AbstractJoinRecordCursorFactory {
+    private final AsOfJoinKeyedFastRecordCursor cursor;
+    private final RecordSink masterKeySink;
+    private final OffheapSink masterSinkTarget = new OffheapSink();
+    private final RecordSink slaveKeySink;
+    private final OffheapSink slaveSinkTarget = new OffheapSink();
+
+    public AsOfJoinFastRecordCursorFactory(
+            CairoConfiguration configuration,
+            RecordMetadata metadata,
+            RecordCursorFactory masterFactory,
+            RecordSink masterKeySink,
+            RecordCursorFactory slaveFactory,
+            RecordSink slaveKeySink,
+            int columnSplit,
+            JoinContext joinContext) {
+        super(metadata, joinContext, masterFactory, slaveFactory);
+        assert slaveFactory.supportsTimeFrameCursor();
+        this.masterKeySink = masterKeySink;
+        this.slaveKeySink = slaveKeySink;
+        this.cursor = new AsOfJoinKeyedFastRecordCursor(
+                columnSplit,
+                NullRecordFactory.getInstance(slaveFactory.getMetadata()),
+                masterFactory.getMetadata().getTimestampIndex(),
+                slaveFactory.getMetadata().getTimestampIndex(),
+                configuration.getSqlAsOfJoinLookAhead()
+        );
+    }
+
+    @Override
+    public boolean followedOrderByAdvice() {
+        return masterFactory.followedOrderByAdvice();
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        RecordCursor masterCursor = masterFactory.getCursor(executionContext);
+        TimeFrameRecordCursor slaveCursor = null;
+        try {
+            slaveCursor = slaveFactory.getTimeFrameCursor(executionContext);
+            cursor.of(masterCursor, slaveCursor);
+            return cursor;
+        } catch (Throwable e) {
+            Misc.free(slaveCursor);
+            Misc.free(masterCursor);
+            throw e;
+        }
+    }
+
+    @Override
+    public int getScanDirection() {
+        return masterFactory.getScanDirection();
+    }
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return false;
+    }
+
+    @Override
+    public void toPlan(PlanSink sink) {
+        sink.type("AsOf Join Fast Scan");
+        sink.attr("condition").val(joinContext);
+        sink.child(masterFactory);
+        sink.child(slaveFactory);
+    }
+
+    @Override
+    protected void _close() {
+        Misc.freeIfCloseable(getMetadata());
+        Misc.free(masterFactory);
+        Misc.free(slaveFactory);
+        Misc.free(masterSinkTarget);
+        Misc.free(slaveSinkTarget);
+    }
+
+    private static class OffheapSink implements RecordSinkSPI, QuietCloseable {
```

Comment:
nit: this class should implement `Mutable` (`clear` instead of `reset`) and `Reopenable` (explicit `reopen` call on `of` cursor call) to keep things consistent with other classes.